js/modules/k6/experimental/csv/module.go

```diff
@@ -0,0 +1,286 @@
+
+	rt := mi.vu.Runtime()
+
+	// 1. Make sure the Sobek object is a fs.File (sobek operation)
+	var fileObj fs.File
+	if err := mi.vu.Runtime().ExportTo(file, &fileObj); err != nil {
+		reject(fmt.Errorf("first argument expected to be a fs.File instance, got %T instead", file))
+		return promise
+	}
+
+	parserOptions := newDefaultParserOptions()
+	if options != nil {
+		var err error
+		parserOptions, err = newParserOptionsFrom(options.ToObject(rt))
+		if err != nil {
+			reject(fmt.Errorf("failed to interpret the provided Parser options; reason: %w", err))
+			return promise
+		}
+	}
+
+	r, err := NewReaderFrom(fileObj.ReadSeekStater, parserOptions)
+	if err != nil {
+		reject(fmt.Errorf("failed to create a new parser; reason: %w", err))
+		return promise
+	}
+
+	go func() {
+		underlyingSharedArrayName := parseSharedArrayNamePrefix + strconv.Itoa(time.Now().Nanosecond())
+
+		// Because we rely on the data module to create the shared array, we need to
+		// make sure that the data module is initialized before we can proceed, and that we don't instantiate
+		// it multiple times.
+		//
+		// As such we hold a single instance of it in the RootModule, and we use it to create the shared array.
+		resolve(mi.RootModule.dataModuleInstance.NewSharedArrayFrom(mi.vu.Runtime(), underlyingSharedArrayName, r))
+	}()
+
+	return promise
+}
+
+// NewParser creates a new CSV parser instance.
+func (mi *ModuleInstance) NewParser(call sobek.ConstructorCall) *sobek.Object {
+	rt := mi.vu.Runtime()
+
+	if mi.vu.State() != nil {
+		common.Throw(rt, errors.New("csv Parser constructor must be called in the init context"))
+	}
+
+	if len(call.Arguments) < 1 || sobek.IsUndefined(call.Argument(0)) {
+		common.Throw(rt, fmt.Errorf("csv Parser constructor takes at least one non-nil source argument"))
+	}
+
+	// 1. Make sure the Sobek object is a fs.File (Sobek operation)
+	var file fs.File
+	if err := mi.vu.Runtime().ExportTo(call.Argument(0), &file); err != nil {
+		common.Throw(
+			mi.vu.Runtime(),
+			fmt.Errorf("first argument expected to be a fs.File instance, got %T instead", call.Argument(0)),
+		)
+	}
+
+	options := newDefaultParserOptions()
+	if len(call.Arguments) == 2 && !sobek.IsUndefined(call.Argument(1)) {
+		var err error
+		options, err = newParserOptionsFrom(call.Argument(1).ToObject(rt))
+		if err != nil {
+			common.Throw(rt, fmt.Errorf("encountered an error while interpreting Parser options; reason: %w", err))
+		}
+	}
+
+	// Instantiate and configure a csv reader using the provided file and options
+	r, err := NewReaderFrom(file.ReadSeekStater, options)
+	if err != nil {
+		common.Throw(rt, fmt.Errorf("failed to create a new parser; reason: %w", err))
+	}
+
+	// Create a new Parser instance
+	parser := Parser{
+		reader:  r,
+		options: options,
+		vu:      mi.vu,
+	}
+
+	return rt.ToValue(&parser).ToObject(rt)
+}
+
+// Next returns the next row in the CSV file.
+func (p *Parser) Next() *sobek.Promise {
```

Comment:
Nit; what happens if `Next` is called after done? Will it resolve to `{ "done": true } again? And `records`? Will be `nil`? 🤔

Reply:
Good question, let me look into it 👀

Reply:
I've added a test for this. When `{done: true}`, the `value` field will be an empty array 🙇🏻